File: best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        # Naive solution O(n * n)
        # max_profit = 0
        # for sell_day in reversed(range(len(prices))):
        #     for buy_day in (range(sell_day)):
        #         potential = prices[sell_day] - prices[buy_day]
        #         if potential > max_profit:
        #             max_profit = potential
        # return max_profit

        # Faster solution O(n)
        max_profit = 0
        if len(prices):
            buy_price = max(prices)
            for price in prices:
                if price < buy_price:
                    buy_price = price
                    potential = 0
                else:
                    potential = price - buy_price
                    if potential > max_profit:
                        max_profit = potential
        return max_profit

File: Best Time to Buy and Sell Stock.js
// Say you have an array for which the ith element is the price of a given stock on day i.

// If you were only permitted to complete at most one transaction (i.e., buy one and sell one share of the stock), design an algorithm to find the maximum profit.

// Note that you cannot sell a stock before you buy one.

/**
 * @param {number[]} prices
 * @return {number}
 */

/* One liner using reduce() */

var maxProfit = function(prices) {
  return prices.reduce(
    (profit, price) => Math.max(profit, Math.max(Math.max(...prices.slice(prices.indexOf(price) + 1)) - price, 0)),
    0
  );
};

/* One liner expanded */

// var maxProfit = function(prices) {
//   return prices.reduce((profit, price) => {
//     var i = prices.indexOf(price);
//     var rest = prices.slice(i + 1);
//     var max = Math.max(...rest);
//     var potential = Math.max(max - price, 0);
//     profit = Math.max(profit, potential);
//     return profit;
//   }, 0);
// };
